### PR TITLE
Request timeout

### PR DESCRIPTION
**Pull request by davimacedo**

### CODE REVIEW

**Comment by davimacedo** [comment]
@TysonAndre what do you think?

File: lib/client.js [hunk 1]
```diff
@@ -128,7 +152,7 @@ module.exports = function (dependencies) {
             resolve(this.write(notification, device, retryCount + 1));
             return;
           } else if (status === 500 && response.reason === "InternalServerError") {
-            this.session.destroy();
+            closeAndDestroy();
```
Comment (TysonAndre) [inline lib/client.js:155]: The closeAndDestroy seems to be declared within an if statement. I'm not sure if it's callable - it may be referring to a global variable instead. You could use the tests in test/client.js to test if this works

File: lib/client.js [hunk 2]
```diff
@@ -20,58 +20,82 @@ module.exports = function (dependencies) {
...
+            logger("No Ping response after " + duration + " ms");
+          }
+          logger("Ping response after " + duration + " ms");
+        });
+      }
+    }, this.config.heartBeat).unref();
   }
 
   Client.prototype.write = function write (notification, device, count) {
+    if (this.isDestroyed) {
```
Comment (TysonAndre) [inline lib/client.js:35]: Probably a useful addition to avoid edge cases

File: lib/client.js [hunk 3]
```diff
@@ -20,58 +20,82 @@ module.exports = function (dependencies) {
...
+      const session = this.session = http2.connect(`https://${this.config.address}`, this.config);
+
+      const destroy = (callback) => {
+        if (this.session === session) {
+          this.session = null;
+        }
+        if (!session.destroyed) {
+          session.destroy();
+        }
+        if (callback) {
```
Comment (TysonAndre) [inline lib/client.js:50]: I think that destroying a session and the session.destroyed flag being set may not be guaranteed to be synchronous

File: lib/client.js [hunk 4]
```diff
@@ -139,7 +163,23 @@ module.exports = function (dependencies) {
           let error = new VError("stream ended unexpectedly");
           resolve({ device, error });
         }
-      })
+      });
+
+      request.setTimeout(this.config.requestTimeout, () => {
```
Comment (TysonAndre) [inline lib/client.js:168]: 1. If the remote server stops sending any TCP/UDP packets back (e.g. no resets, no acknowledgements, no close notifications), then I'm concerned that this could resolve promises but never close the http/2 session if requests are sent but no requests are getting responses. (on('error') would be emitted in response to a packet)

   (I assume the failure mode I've been seeing is similar to this, which is what I wrote my PR to address, including tracking the most recent successsful response time across *all* requests, to work even if bandwidth was limited)
2. If this doesn't close the connection, it would leak memory.

File: lib/client.js [hunk 5]
```diff
@@ -139,7 +163,23 @@ module.exports = function (dependencies) {
...
+        if (logger.enabled) {
+          logger('Request timeout');
+        }
+
+        resolve({ device, error: new VError("apn write timeout") });
+      });
+
+      request.on("aborted", () => {
```
Comment (TysonAndre) [inline lib/client.js:176]: Probably useful to add for diagnostics - I assume end would get emitted after that

Comment (davimacedo) [reply lib/client.js:176]: Yes. That's the idea.

File: lib/client.js (hunk 2 above)
Comment (davimacedo) [reply lib/client.js:35]: Yes. I think it is.

File: lib/client.js (hunk 3 above)
Comment (davimacedo) [reply lib/client.js:50]: This verification is only to avoid calling destroy function twice. According to the documentation, the session is immediately destroyed and since there is no callback for this function I believe it is synchronous.
https://nodejs.org/api/http2.html#http2_http2session_destroy_error_code

File: lib/client.js (hunk 1 above)
Comment (davimacedo) [reply lib/client.js:155]: You're right. I will fix that.

Comment (davimacedo) [reply lib/client.js]: done

File: lib/client.js (hunk 4 above)
Comment (davimacedo) [reply lib/client.js:180]: 1 - So what's your suggestion here? Close and destroy the session to create a new one? I'm not sure if it is necessary since the server can timeout for this request but can work for the next one. I did a stress test and saw this happening. One request timed out but the following ones went through. I believe that if something wrong goes with the whole session, one of the other events will be fired.
2 - You're right. I will close the request.

Comment (davimacedo) [reply lib/client.js:181]: 2 - done

Comment (TysonAndre) [reply lib/client.js:181]: > 1 - So what's your suggestion here? Close and destroy the session to create a new one? I'm not sure if it is necessary since the server can timeout for this request but can work for the next one. I did a stress test and saw this happening. One request timed out but the following ones went through. I believe that if something wrong goes with the whole session, one of the other events will be fired.

The suggestion was to track _lastSuccessfulApnResponseMillis - if there are active requests, only give up on the connection if the most recently received successful response was received earlier than `now - timeout`.

I can't say more without being sure of what the stress test was doing and what it was tested against

**Comment by TysonAndre** [comment]
I'm not sure if you've seen this, but https://github.com/parse-community/node-apn/pull/27/files#diff-7e3bd8e7059a21bfd255649cbfc84b33 added various integration tests for the client properly handling some of the edge cases. You may want to copy those to make sure that the code works as intended on the supported node versions.

File: lib/client.js [hunk 6]
```diff
@@ -14,64 +14,101 @@ module.exports = function (dependencies) {
...
-                     logger("No Ping response after " + duration + " ms");
-                 }
-                 logger("Ping response after " + duration + " ms");
-             })
-         }
-     }, this.config.heartBeat).unref();
+      if (this.session && !this.session.closed && !this.session.destroyed && !this.isDestroyed) {
+        this.session.ping((error, duration) => {
+          if (error) {
+            logger("No Ping response after " + duration + " ms");
```
Comment (TysonAndre) [inline lib/client.js:27]: nyc: should return instead of falling through. It may also be useful to log the error message

File: lib/client.js [hunk 7]
```diff
@@ -139,7 +176,25 @@ module.exports = function (dependencies) {
...
         }
-      })
+      });
+
+      request.setTimeout(this.config.requestTimeout, () => {
+        if (logger.enabled) {
+          logger('Request timeout');
+        }
+
+        request.close(NGHTTP2_CANCEL);
```
Comment (TysonAndre) [inline lib/client.js:186]: Closing the request sends a message indicating that the **request** should be cancelled by the http2 library over the session, but does not close the session itself - new requests will continue to be sent over the session, even if the remote server is no longer sending any response to packets such as acknowledgements or indicating that the connection is invalid/terminated

https://en.wikipedia.org/wiki/HTTP/2#Goals - i.e. the `session` is a TCP socket, and the `request`s are multiplexed by the http2 library over a single TCP session.

**Comment by TysonAndre** [comment]
Looking at it again,
- The issues I remember with APNs silently terminating TCP connections were with the binary protocol, not the HTTP/2 protocol. It's possible but not certain HTTP/2 has similar issues.
- I'm still learning about http/2 in node

The node library seems fairly low level and seems to be leaving applications to detect dead connections and reconnect.

The golang library seems more mature (This is a completely different language, but looking at other languages may help in thinking up robust solutions). In it, it internally maintains a connection pool. In order to detect dead connections, it sends a ping frame and checks if any ping response(pong) was sent within 15 seconds of that ping. https://github.com/golang/go/blob/f8d80977b784fd4879963e61dc9fca1fc9bf2193/src/net/http/h2_bundle.go#L7202-L7214

- This helps distinguish between disconnection in the transport and application delays, I guess 
- https://godoc.org/golang.org/x/net/http2#Transport has PingTimeout default to 15 seconds

(In my experience, https://github.com/uniqush/uniqush-push/ (https://github.com/uniqush/uniqush-push/blob/master/srv/apns/http_api/processor.go) has been relatively stable and these issues have been new when working on a project using node-apn, but that may just be a consequence of other factors, such as golang's http2 library vs node's. It could also be some other factor like the OS or the system load being lower)

File: lib/client.js (hunk 7 above)
Comment (TysonAndre) [reply lib/client.js:186]: I know - I'm confused about what session.setTimeout is actually doing.

E.g. What happens if I repeatedly call setTimeout every 20 milliseconds - It appears at a glance as if the old timeout gets cleared and replaced with the new timeout, but I may be mistaken.
- If the old timeout does actually get cleared, that would be a problem if send() keeps getting called

**Comment by isidai** [comment]
I’m currently using @parse-community/node-apn version 3.1.0. I came here because I also encountered an issue with the Promise not resolving after a timeout.

I enabled debug logging and got the following logs when I was having this problem

```
apn Request error: Error [ERR_HTTP2_STREAM_CANCEL]: The pending stream has been canceled (caused by: connect ETIMEDOUT xxx.xxx.xxx.xxx:443)
apn Request ended with status null and responseData: 
apn Session error: Error: connect ETIMEDOUT xxx.xxx.xxx.xxx:443
apn Session closed
```

When looking into this pull request, it's not really the root-cause of the issue, but I did notice a few problems.

1. The duration passed in as an argument to this.session.ping is undefined in the event of an error, so should there be an error, the duration is not properly set.

2. When an error is caught, the ping response was found to be undefined.
"No ping response after undefined ms"
"Ping response after undefined ms"

There also didn't seem to be any recovery process from any errors that could occur during the health check.
https://github.com/parse-community/node-apn/blob/master/lib/provider.js#L45

3. The shutdown function in provider.js does not pass any arguments.
The client.js shutdown allows you to pass a callback, but you can't pass a callback from the provider.

4. There's also a discrepency with the "Note" in the readme (link 1) stemming from (link 2)
Link 1: https://github.com/parse-community/node-apn/blob/master/doc/provider.markdown#connectionshutdown
Link 2: https://github.com/parse-community/node-apn/blob/7d403855adec5e364707db3d21df2cd13a4c47ef/lib/client.js#L71

**Comment by davimacedo** [comment]
Hi guys.

I believe that there is a lot to be improved here and the idea of managing a pool of connections would be the best. But, at least for my case, this PR looks correctly addressing the issue with stuck pushes. Time to time the pushes were stuck and my understanding now is that it happens because if for some reason Apple do not respond to the request, the request is never closed and the promise never resolved. I tested this new PR and the problem stopped to happen. I've sent about 1M push notifications. I also enabled the debug logs while sending the pushes and I've noticed that some of them timed out but the process resumed smoothly for the next requests.

I'm only afraid to start addressing a lot of different issues (most of them regarding the legacy code) in this PR and we turn out taking a lot of time to solve the main issue.

So my suggestion is: we list the no-show items we should fix before merging this PR,  fix them, and address the other ones in future PRs. What do you think we should fix for merging this PR?

Also please feel free to either commit suggested changes in this PR or open new PRs with other issues.

It would also be good if few of you could test this branch in your environment as well.

File: lib/client.js (hunk 4 above)
Comment (davimacedo) [reply lib/client.js:181]: I'm not sure if we should do that in this PR. It brings an extra complexity to the code. I believe we could address on a separate PR if we continue having issues.

File: lib/client.js (hunk 6 above)
Comment (davimacedo) [reply lib/client.js:27]: Yes. Make sense. Let's open another PR for this?

File: lib/client.js (hunk 7 above)
Comment (davimacedo) [reply lib/client.js:186]: I believe we should go with the request.setTimeout. In my tests I've seen a request timing out but the next one in the same session successfully completed. I believe that if something happen to fail the whole session, we will have the error event being fired and the connection being renewed.

Comment (petitchevalroux) [reply lib/client.js:186]: @TysonAndre I tested calling session.setTimeout repeatedly (see : https://github.com/petitchevalroux/node/blob/http2-timeout/http2-timeout/session-timeout.js). I get the following output : 
```
session received
Wed Oct 07 2020 17:09:19 GMT+0200 (GMT+02:00)
session timeout 1
Wed Oct 07 2020 17:09:24 GMT+0200 (GMT+02:00)
session timeout 2
```
As there is no multiple "session timeout X" between date log, it seems to me that you are right saying that the old timeout is replaced every time. 

[Here](https://github.com/petitchevalroux/node/blob/http2-timeout/http2-timeout/request-timeout.js) I tested successfully (as @davimacedo) if a request can suceeed after a previous one timed out using the same session.

To me session.setTimeout handler is just trigger if nothing happen on the session (as the doc says). We could use it as a new check when sending a request but actually it does not add much compared to the setTimeout of the request. If the tcp socket time out, the request will too.

Sorry, I am now convinced that it is not necessary.

**Comment by davimacedo** [comment]
@dplewis @funkenstrahlen would you mind to take a look here as well? I wanted to have it merged for the next version of parse server.

File: lib/client.js (hunk 7 above)
Comment (TysonAndre) [reply lib/client.js:186]: > Here I tested successfully (as @davimacedo) if a request can suceeed after a previous one timed out using the same session.

You should be aware that https://nodejs.org/api/http2.html#http2_request_settimeout_msecs_callback has a different behavior if a callback or an event handler is declared for 'timeout' - It destroys the Http2Stream if there is no callback/event handler.

To actually destroy the request, you may need to call request.destroy - I think it might slowly leak memory otherwise, depending on what APNs is doing, but I'm not familiar with the implementation

> Each instance of the Http2Stream class represents a bidirectional HTTP/2 communications stream over an Http2Session instance. Any single Http2Session may have up to `2^31-1` Http2Stream instances over its lifetime.



> If no 'timeout' listener is added to the request, the response, or the server, then Http2Streams are destroyed when they time out. If a handler is assigned to the request, the response, or the server's 'timeout' events, timed out sockets must be handled explicitly.

Comment (TysonAndre) [reply lib/client.js:186]: Anyway, it's possible that this might fix this - I wouldn't know until we try that, and your test results seem promising if the connection is still working.

It seems like https://developers.google.com/web/fundamentals/performance/http2#flow_control might help - I was initially concerned that the  NGHTTP2_CANCEL frame might have just been more data sent over the TCP socket, and if the tcp send buffer was full, APNs wouldn't get the frame, but if there's additional flow control, cancellation might work properly assuming it's more urgent than new requests.

I wonder if the send issues might be specific to OS versions or node versions or the hardware we're running on, but hopefully it's a reproducible bugs (e.g. only some users run into that and others don't)

File: lib/client.js [hunk 8]
```diff
@@ -14,64 +14,101 @@ module.exports = function (dependencies) {
...
+      session = this.session;
+    }
+    if (session) {
+      if (this.session === session) {
+        this.session = null;
+      }
+      if (!session.closed) {
+        session.close(() => this.destroy(session, callback));
+      } else {
+        this.destroy(session, callback)
```
Comment (TysonAndre) [inline lib/client.js:63]:
```suggestion
        this.destroySession(session, callback)
```

```
  Client.prototype.destroySession = function (session, callback) {
```

File: lib/client.js [hunk 9]
```diff
@@ -14,64 +14,101 @@ module.exports = function (dependencies) {
...
+  Client.prototype.closeAndDestroySession = function (session, callback) {
+    if (!session) {
+      session = this.session;
+    }
+    if (session) {
+      if (this.session === session) {
+        this.session = null;
+      }
+      if (!session.closed) {
+        session.close(() => this.destroy(session, callback));
```
Comment (TysonAndre) [inline lib/client.js:61]:
```suggestion
        session.close(() => this.destroySession(session, callback));
```

**Comment by TysonAndre** [comment]
> I'm not sure if you've seen this, but https://github.com/parse-community/node-apn/pull/27/files#diff-7e3bd8e7059a21bfd255649cbfc84b33 added various integration tests for the client properly handling some of the edge cases. You may want to copy those to make sure that the code works as intended on the supported node versions.

The test case should have caught the `this.destroy` bug if it was added. The test cases that seem unrealistic such as pathologically slow responses could be removed or adjusted for the test expectations (e.g. expect all promises to be resolved and some minimum/maximum number to get successful responses)

File: lib/client.js (hunk 8 above)
Comment (davimacedo) [reply lib/client.js]: Thanks for pointing out. Fixed.

File: lib/client.js (hunk 9 above)
Comment (davimacedo) [reply lib/client.js]: Thanks for pointing out. Fixed.

File: lib/client.js (hunk 6 above)
Comment (davimacedo) [reply lib/client.js]: done

File: lib/client.js (hunk 7 above)
Comment (davimacedo) [reply lib/client.js:195]: I actually used the same code that I found in node.js docs and that's why I believe it has more chances to be the right way to go: https://nodejs.org/api/http2.html#http2_http2stream_settimeout_msecs_callback

It successfully worked in my tests and I believe with should go with this and maybe change later if some new issues arise.

**Comment by davimacedo** [comment]
> > I'm not sure if you've seen this, but https://github.com/parse-community/node-apn/pull/27/files#diff-7e3bd8e7059a21bfd255649cbfc84b33 added various integration tests for the client properly handling some of the edge cases. You may want to copy those to make sure that the code works as intended on the supported node versions.
> 
> The test case should have caught the `this.destroy` bug if it was added. The test cases that seem unrealistic such as pathologically slow responses could be removed or adjusted for the test expectations (e.g. expect all promises to be resolved and some minimum/maximum number to get successful responses)

I've included the tests. I've just removed the one that tries to simulate a loaded server.

File: lib/client.js [hunk 10]
```diff
@@ -14,64 +14,102 @@ module.exports = function (dependencies) {
...
+            logger("No Ping response after " + duration + " ms with error:" + error.message);
+            return;
+          }
+          logger("Ping response after " + duration + " ms");
+        });
+      }
+    }, this.config.heartBeat).unref();
+  }
+
+  Client.prototype.destroySession = function (session, callback) {
```
Comment (TysonAndre) [inline lib/client.js:36]: Could document here and in closeAndDestroySession that session should be set except when destroying the client

**Review by TysonAndre — COMMENTED** [review]
Looks good overall.

Would prefer weakening the test of an overloaded server (e.g. all promises resolve) over removing the test due to issues (in the tested code?) but not a blocker since this is a large PR

**Comment by petitchevalroux** [comment]
Just to give you feedback. We have been using this PR in production since last Thursday with success. The hanging problem seems to be resolved. Thanks for your work.

**Comment by davimacedo** [comment]
> Just to give you feedback. We have been using this PR in production since last Thursday with success. The hanging problem seems to be resolved. Thanks for your work.

It is good to know!

Guys, if no other concern arises, I am ready to mode forward in merging this PR. Please let me know.

File: lib/client.js [hunk 11]
```diff
@@ -112,34 +152,60 @@ module.exports = function (dependencies) {
...
+        if (logger.enabled) {
+          logger('Request aborted');
         }
-      })
+
+        resolve({ device, error: new VError("apn write aborted") });
+      });
 
       request.on("error", (error) => {
         if (logger.enabled) {
```
Comment (TysonAndre) [inline lib/client.js:211]: I noticed an error when running typescript to see if there were similar issues to those I had already seen.

**This may cause the promise to hang if an apn write fails - this will probably throw a ReferenceError due to the undefined variable**

```patch
         if (typeof error === "string") {
-          error = new VError("apn write failed: %s", err);
++          error = new VError("apn write failed: %s", error);
```

I noticed this using typescript (some options in tsconfig.json are unnecessary/inapplicable)

```js
// tsconfig.json
{
  "compilerOptions": {
    "target": "es6",
    "module": "commonjs",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "outDir": ".built",
    "allowJs": true,
    "checkJs": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true
  },
  "exclude": ["index.d.ts", ".built"]
}
```

Comment (TysonAndre) [reply lib/client.js:211]: https://nodejs.org/api/http2.html#http2_class_http2stream claims that the 'error' should be an instance of `Error`, though, so hopefully not the cause of any issues.

LGTM otherwise

Comment (TysonAndre) [reply lib/client.js:211]: https://github.com/nodejs/node/blob/master/lib/internal/http2/core.js seems like it doesn't enforce that the error is an object, e.g. if the caller of destroy() passes in a string instead of an Error object, but it doesn't seem as if this is the reason for node-apn crashing in practice

Comment (TysonAndre) [reply lib/client.js:211]: https://github.com/TysonAndre/node-apn/commit/60b760d896d7712c64d88146f062dd4e02df83f0 fixes that and some nits on trailing whitespace

File: lib/client.js [hunk 12]
```diff
@@ -154,22 +220,29 @@ module.exports = function (dependencies) {
         resolve({ device, error });
       });
 
+      if (logger.enabled) {
```
Comment (TysonAndre) [inline lib/client.js:223]: This block is wrong and should be removed - I believe this leads to a memory leak. The event handler for session were already handled elsewhere

A given session can have an arbitrary number of requests.

This is the handler for a request - so this ends up registering an arbitrary number of callbacks for frameError

File: lib/client.js [hunk 13]
```diff
@@ -154,22 +220,29 @@ module.exports = function (dependencies) {
...
+          logger(`Request frame error: (frameType: ${frameType}, errorCode ${errorCode}, streamId: ${streamId})`);
+        });
+      }
+
       request.end();
     });
   };
 
   Client.prototype.shutdown = function shutdown(callback) {
+    if (logger.enabled) {
```
Comment (TysonAndre) [inline lib/client.js:238]: nit: Could return early (and call callback?) if this.isDestroyed is already true (i.e. something called shutdown earlier)?

Out of scope for this PR, though, and not a concern for well written programs

File: lib/client.js [hunk 14]
```diff
@@ -154,22 +220,29 @@ module.exports = function (dependencies) {
...
+      logger('Called client.shutdown()');
+    }
+    this.isDestroyed = true;
     if (this.healthCheckInterval) {
       clearInterval(this.healthCheckInterval);
```
Comment (TysonAndre) [inline lib/client.js:243]: nit: could set healthCheckInterval to null after clearing the interval

**Review by TysonAndre — CHANGES_REQUESTED** [review]
LGTM other than the event listener leak when the environment variable `DEBUG=apn` is set

File: lib/client.js (hunk 12 above)
Comment (rileyweber) [reply lib/client.js:223]: This seems to definitely be the case. We were seeing memory usage increasing linearly in our staging environment, but after removing this block memory is stable again

File: lib/client.js [hunk 15]
```diff
@@ -112,34 +152,60 @@ module.exports = function (dependencies) {
 
     request.write(notification.body);
 
-    return new Promise ( resolve => {
+    return new Promise ( resolve => {      
       request.on("end", () => {
+        try {
+          if (logger.enabled) {
+            logger(`Request ended with status ${status} and responseData: ${responseData}`);
```
Comment (TysonAndre) [inline lib/client.js:159]: This results in the error message `apn Request ended with status null and responseData: ` when a request times out or is aborted, which may confuse people reading the logs.

Setting `status = "(timed out)"` in the 'timeout' handler (and possibly something similar for aborted, etc) is one way this error message could be made more readable, but not a blocker

File: lib/client.js (hunk 12 above)
Comment (TysonAndre) [reply lib/client.js:223]: So far, this seems to be working reasonably well in production, using the suggested patches as https://github.com/TysonAndre/node-apn/commits/ae8aa835b1689604047b78bee567bdfe706ff197 (with DEBUG=apn)

File: lib/client.js (hunk 11 above)
Comment (davimacedo) [reply lib/client.js:211]: done

File: lib/client.js (hunk 12 above)
Comment (davimacedo) [reply lib/client.js:223]: Fixed. it should be `request` and not `this.session`

File: lib/client.js (hunk 13 above)
Comment (davimacedo) [reply lib/client.js:244]: done

File: lib/client.js (hunk 14 above)
Comment (davimacedo) [reply lib/client.js:249]: done

**Comment by TysonAndre** [comment]
The latest 5 commits seem reasonable; I haven't tested them but I don't expect any issues

File: lib/client.js [hunk 16]
```diff
@@ -114,62 +154,102 @@ module.exports = function (dependencies) {
...
+          error = new VError("apn write failed: %s", error);
         } else {
           error = new VError(error, "apn write failed");
         }
         resolve({ device, error });
       });
 
+      if (logger.enabled) {
+        request.on("close", () => {
+          logger("Request closed");
```
Comment (TysonAndre) [inline lib/client.js:225]: nit: This seems harmless but excessively verbose - It seems at a glance like any request would emit both 'end' and 'close' whether it succeeded or failed. So we'd already be indicating this with logs such as ` Request ended with status null and responseData`

e.g. running the unit tests with DEBUG=apn

```
  apn Session connected +0ms
  apn Request ended with status null and responseData:  +12ms
  apn Request closed +0ms
  apn Request ended with status null and responseData:  +0ms
  apn Request closed +0ms
  apn Request ended with status null and responseData:  +0ms
  apn Request closed +0ms
  apn GOAWAY received: (errorCode 0, lastStreamId: 7, opaqueData: undefined) +0ms
  apn Request ended with status null and responseData:  +1ms
  apn Request closed +0ms
  apn Session closed +0ms

```

Comment (davimacedo) [reply lib/client.js]: ok. agree. I've just removed.

File: lib/client.js [hunk 17]
```diff
@@ -114,62 +154,98 @@ module.exports = function (dependencies) {
...
+          } else if (responseData !== "") {
+            const response = JSON.parse(responseData);
+
+            if (status === 403 && response.reason === "ExpiredProviderToken" && retryCount < 2) {
+              this.config.token.regenerate(tokenGeneration);
+              resolve(this.write(notification, device, retryCount + 1));
+              return;
+            } else if (status === 500 && response.reason === "InternalServerError") {
+              this.closeAndDestroySession();
+              let error = new VError("Error 500, stream ended unexpectedly");
```
Comment (TysonAndre) [inline lib/client.js:173]: https://travis-ci.com/github/parse-community/node-apn/jobs/399143362 - travis failure seems spurious, could add a generic helper such as assertErrorIsStreamEndedUnexpectedly to expect either error format

Comment (davimacedo) [reply lib/client.js:173]: I've amended. What do you think?

File: lib/client.js (hunk 15 above)
Comment (TysonAndre) [reply lib/client.js:159]: This is what I've seen in the actual logs for push attempts to APNs

This can also be reproduced by running `DEBUG=apn node_modules/.bin/mocha test/client.js`

```
  apn Request closed +1ms
  apn Called client.shutdown() +0ms
  apn Session closed +1ms
  apn Session connected +3ms
  apn Request timeout +502ms
  apn Request ended with status null and responseData:  +3ms
  apn Request closed +0ms
  apn Request timeout +501ms
  apn Request timeout +0ms
  apn Request ended with status null and responseData:  +1ms
  apn Request closed +0ms
  apn Request timeout +0ms
  apn Request ended with status null and responseData:  +0ms
  apn Request closed +1ms
  apn Request timeout +0ms
  apn Request ended with status null and responseData:  +0ms
  apn Request closed +0ms
    ✓ Handles APNs timeouts (1011ms)

```

File: lib/client.js (hunk 17 above)
Comment (TysonAndre) [reply lib/client.js:173]: The amended changes look good to me, I don't have any new comments on the latest commits to the PR

File: lib/client.js (hunk 15 above)
Comment (davimacedo) [reply lib/client.js:163]: ok. just sent an additional commit. could you please check?

**Comment by TysonAndre** [comment]
LGTM for the latest changes, logs seem reasonable for the test output with DEBUG=apn for common use cases

**Comment by davimacedo** [comment]
Nice. Thank you so much for all your help! And let's continue addressing the other ideas you've raised. Feel free to open new PRs as well.
